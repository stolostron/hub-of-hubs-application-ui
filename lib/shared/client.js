/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2017, 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
// Copyright Contributors to the Open Cluster Management project
'use strict'

/*
 * A boolean module that is true on the client, false on the server. This is for code that needs to
 * do different things on the server vs the client.
 */

module.exports = typeof window !== 'undefined'
