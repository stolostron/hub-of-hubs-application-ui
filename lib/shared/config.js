/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
// Copyright (c) 2020 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
var requireServer = require('./require-server'),
    nconf = requireServer('nconf')

var WHITELIST = [
  'API_SERVER_URL',
  'contextPath',
  'hcmUiApiUrl',
  'searchApiUrl',
  'searchQueryLimit',
  'overview',
  'featureFlags: filters',
  'featureFlags:dashboardLiveUpdates',
  'featureFlags:dashboardRefreshInterval',
  'featureFlags:fullDashboard',
  'featureFlags:liveUpdates',
  'featureFlags:liveUpdatesPollInterval',
  'featureFlags:search',
  'featureFlags:clusterFeature',
  'feature_searchRelated',
  'featureFlags:enableSegment',
  'featureFlags:segmentKey'
]

var config = {}

if (nconf) {
  WHITELIST.forEach(i => (config[i] = nconf.get(i)))
  config.env = process.env.NODE_ENV
} else {
  const configElement = document.getElementById('config')
  config = (configElement && JSON.parse(configElement.textContent)) || {}
}

module.exports = config
