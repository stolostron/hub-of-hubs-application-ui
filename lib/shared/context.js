/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2017, 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
// Copyright (c) 2020 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
'use strict'

var requireServer = require('./require-server'),
    locale = requireServer('./locale')
/*
 * A function that provides context information for a given page or request. e.g. who the logged in user is.
 * Uses the session on the server, and the context payload script on the client.
 */

module.exports = function(req) {
  if (req) {
    return {
      locale: locale(req)
    }
  }
  return JSON.parse(document.getElementById('context').textContent)
}
