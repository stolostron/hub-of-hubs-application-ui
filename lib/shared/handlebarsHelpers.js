// Copyright (c) 2020 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
'use strict'

const i18n = require('node-i18n-util'),
      path = require('path')

module.exports = function(href, locale) {
  if (href && locale) {
    const relativePath = `../../${href}` // load the properties
    require(relativePath)
    // resolve the bundle locale being used
    const bundlePath = path.join(__dirname, relativePath),
          localeTemp = i18n._resolveBundle(bundlePath, locale).locale,
          basePath = href.substring(0, href.indexOf('.properties'))
    href =
      basePath + (localeTemp ? '_' + localeTemp.replace('-', '_') : '') + '.js'
    return href
  }
  return ''
}
