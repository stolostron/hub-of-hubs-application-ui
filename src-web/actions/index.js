/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2017, 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
// Copyright (c) 2020 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
/*
 * action types
 *
 * action name: <NOUN>_<VERB>
 */

export { applicationPageResources } from '../reducers/application'
export * from '../reducers/reducerAppDeployments'
export * from '../reducers/reducerAppOverview'

export const CLEAR_REQUEST_STATUS = 'CLEAR_REQUEST_STATUS'

export const APPLICATION_CREATE_CLEAR_STATUS =
  'APPLICATION_CREATE_CLEAR_STATUS'
export const APPLICATION_CREATE_IN_PROGRESS = 'APPLICATION_CREATE_IN_PROGRESS'
export const APPLICATION_CREATE_FAILURE = 'APPLICATION_CREATE_FAILURE'
export const APPLICATION_CREATE_SUCCESS = 'APPLICATION_CREATE_SUCCESS'

export const DEL_RECEIVE_FAILURE = 'DEL_RECEIVE_FAILURE'
export const DEL_RECEIVE_SUCCESS = 'DEL_RECEIVE_SUCCESS'
export const DEL_RECEIVE_SUCCESS_FINISHED = 'DEL_RECEIVE_SUCCESS_FINISHED'

export const DEL_REQUEST = 'DEL_REQUEST'

export const MODAL_UPDATE = 'MODAL_UPDATE'

export const POST_RECEIVE_FAILURE = 'POST_RECEIVE_FAILURE'
export const POST_RECEIVE_SUCCESS = 'POST_RECEIVE_SUCCESS'
export const POST_REQUEST = 'POST_REQUEST'
export const PUT_RECEIVE_FAILURE = 'PUT_RECEIVE_FAILURE'
export const PUT_RECEIVE_SUCCESS = 'PUT_RECEIVE_SUCCESS'
export const PUT_REQUEST = 'PUT_REQUEST'

export const RESOURCES_FETCH_REQUEST_LOADING = 'RESOURCE_FETCH_REQUEST_LOADING'
export const RESOURCES_FETCH_REQUEST_SUCCESS =
  'RESOURCES_FETCH_REQUEST_SUCCESS'
export const RESOURCE_ADD = 'RESOURCE_ADD'
export const RESOURCE_DELETE = 'RESOURCE_DELETE'
export const RESOURCE_MODIFY = 'RESOURCE_MODIFY'
export const RESOURCE_MUTATE = 'RESOURCE_MUTATE'
export const RESOURCE_MUTATE_FINISHED = 'RESOURCE_MUTATE_FINISHED'
export const RESOURCE_MUTATE_FAILURE = 'RESOURCE_MUTATE_FAILURE'
export const RESOURCE_MUTATE_SUCCESS = 'RESOURCE_MUTATE_SUCCESS'
export const RESOURCE_RECEIVE_FAILURE = 'RESOURCE_RECEIVE_FAILURE'
export const RESOURCE_RECEIVE_NOT_FOUND = 'RESOURCE_RECEIVE_NOT_FOUND'
export const RESOURCE_RECEIVE_SUCCESS = 'RESOURCE_RECEIVE_SUCCESS'
export const RESOURCE_RECEIVE_CLUSTER_OFFLINE =
  'RESOURCE_RECEIVE_CLUSTER_OFFLINE'
export const RESOURCE_DETAILS_RECEIVE_SUCCESS =
  'RESOURCE_DETAILS_RECEIVE_SUCCESS'
export const RESOURCE_FILTERS_RECEIVE_SUCCESS =
  'RESOURCE_FILTERS_RECEIVE_SUCCESS'
export const RESOURCE_FILTERS_UPDATE = 'RESOURCE_FILTERS_UPDATE'
export const RESOURCE_REQUEST = 'RESOURCE_REQUEST'
export const RESOURCE_DETAILS_REQUEST = 'RESOURCE_DETAILS_REQUEST'
export const RESOURCE_RESET = 'RESOURCE_RESET'

export const ROLE_RECEIVE_FAILURE = 'ROLE_RECEIVE_FAILURE'
export const ROLE_RECEIVE_SUCCESS = 'ROLE_RECEIVE_SUCCESS'

export const SECONDARY_HEADER_UPDATE = 'SECONDARY_HEADER_UPDATE'

export const TABLE_PAGE_CHANGE = 'TABLE_PAGE_CHANGE'
export const TABLE_SEARCH = 'TABLE_SEARCH'
export const TABLE_SORT = 'TABLE_SORT'

export const TOPOLOGY_NAME_SEARCH = 'TOPOLOGY_NAME_SEARCH'
export const TOPOLOGY_RESTORE_SAVED_FILTERS = 'TOPOLOGY_RESTORE_SAVED_FILTERS'
export const TOPOLOGY_SET_ACTIVE_FILTERS = 'TOPOLOGY_SET_ACTIVE_FILTERS'
export const TOPOLOGY_FILTERS_UPDATE = 'TOPOLOGY_FILTERS_UPDATE'
export const TOPOLOGY_FILTERS_REQUEST = 'TOPOLOGY_FILTERS_REQUEST'
export const TOPOLOGY_FILTERS_RECEIVE_ERROR = 'TOPOLOGY_FILTERS_RECEIVE_ERROR'
export const TOPOLOGY_FILTERS_RECEIVE_SUCCESS =
  'TOPOLOGY_FILTERS_RECEIVE_SUCCESS'

export const DIAGRAM_RESTORE_FILTERS = 'DIAGRAM_RESTORE_FILTERS'
export const DIAGRAM_SAVE_FILTERS = 'DIAGRAM_SAVE_FILTERS'

export const UICONFIG_RECEIVE_SUCCESS = 'UICONFIG_RECEIVE_SUCCESS'

export const UPDATE_USERPREFERENCES = 'UPDATE_USERPREFERENCES'

export const USERPREFERENCES_RECEIVE_FAILURE = 'USERPRFERENCES_RECEIVE_FAILURE'
export const USERPREFERENCES_RECEIVE_SUCCESS = 'USERPRFERENCES_RECEIVE_SUCCESS'

export const USER_LOGIN_RECEIVE_SUCCESS = 'USER_LOGIN_RECEIVE_SUCCESS'
export const USER_LOGOUT_RECEIVE_FAILURE = 'USER_LOGOUT_RECEIVE_FAILURE'
export const USER_LOGOUT_RECEIVE_SUCCESS = 'USER_LOGOUT_RECEIVE_SUCCESS'
export const USER_LOGOUT_REQUEST = 'USER_LOGOUT_REQUEST'

export const RESOURCE_FORCE_RELOAD = 'RESOURCE_FORCE_RELOAD'
export const RESOURCE_FORCE_RELOAD_FINISHED = 'RESOURCE_FORCE_RELOAD_FINISHED'

/*
 * other constants specific to particular actions
 */

export const VisibilityFilters = {
  SHOW_ALL: 'SHOW_ALL',
  SHOW_COMPLETED: 'SHOW_COMPLETED',
  SHOW_ACTIVE: 'SHOW_ACTIVE'
}

export const USER_LOGIN_STATUS = {
  LOGGED_IN: 'LOGGED_IN',
  LOGGED_OUT: 'LOGGED_OUT'
}

export const REQUEST_STATUS = {
  INCEPTION: 'INCEPTION',
  DONE: 'DONE',
  IN_PROGRESS: 'IN_PROGRESS',
  ERROR: 'ERROR',
  NOT_FOUND: 'NOT_FOUND',
  CLUSTER_OFFLINE: 'CLUSTER_OFFLINE'
}

export const PAGE_SIZES = {
  DEFAULT: 20,
  VALUES: [5, 10, 20, 50, 75, 100]
}

export const SORT_DIRECTION_ASCENDING = 'asc'
export const SORT_DIRECTION_DESCENDING = 'desc'
