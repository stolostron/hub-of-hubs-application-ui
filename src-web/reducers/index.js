/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2017, 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
// Copyright (c) 2020 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
/*
 *
 * Things you should never do inside a reducer:
 *
 * - Mutate its arguments
 * - Perform side effects like API calls and routing transitions
 * - Call non-pure functions, e.g. Date.now() or Math.random()
 *
 * Reducers must be deterministic pure functions.  Given the same arguments, it should calculate the next state and return it.
 * No surprises. No side effects. No API calls. No mutations. Just a calculation.
 *
 * Selectors should sit along side reducers.
 */

import { createResourceReducer, resourceReducerFunction } from './common'
import { RESOURCE_TYPES } from '../../lib/shared/constants'
import _ from 'lodash'

export { user, loggedIn } from './user'
export { secondaryHeader } from './common'

export { uiconfig } from './uiconfig'
export { role } from './role'

export { modal } from './modal'

export { applicationPageResources } from './application'
export { AppDeployments } from './reducerAppDeployments'
export { AppOverview } from './reducerAppOverview'

const keyFields = ['name', 'list']

function predicate(resourceType, action) {
  if (
    _.isEqual(
      _.pick(resourceType, keyFields),
      _.pick(action.resourceType, keyFields)
    )
  ) {
    return true
  }
  return _.find(_.values(resourceType), type => {
    if (typeof type === 'string') {
      return type.indexOf(action.resourceType) > -1
    }
    return false
  })
}

// the exported function name must match the resourceType value
export const HCMApplicationList = createResourceReducer(
  resourceReducerFunction,
  predicate.bind(null, RESOURCE_TYPES.HCM_APPLICATIONS)
)

// the exported function name must match the resourceType value
export const QueryApplicationList = createResourceReducer(
  resourceReducerFunction,
  predicate.bind(null, RESOURCE_TYPES.QUERY_APPLICATIONS)
)

export const QueryApplicationsetList = createResourceReducer(
  resourceReducerFunction,
  predicate.bind(null, RESOURCE_TYPES.QUERY_APPLICATIONSET)
)

export const QuerySubscriptionList = createResourceReducer(
  resourceReducerFunction,
  predicate.bind(null, RESOURCE_TYPES.QUERY_SUBSCRIPTIONS)
)

export const QueryPlacementRuleList = createResourceReducer(
  resourceReducerFunction,
  predicate.bind(null, RESOURCE_TYPES.QUERY_PLACEMENTRULES)
)

export const QueryChannelList = createResourceReducer(
  resourceReducerFunction,
  predicate.bind(null, RESOURCE_TYPES.QUERY_CHANNELS)
)

export const HCMChannelList = createResourceReducer(
  resourceReducerFunction,
  predicate.bind(null, RESOURCE_TYPES.HCM_CHANNELS)
)

export const HCMSubscriptionList = createResourceReducer(
  resourceReducerFunction,
  predicate.bind(null, RESOURCE_TYPES.HCM_SUBSCRIPTIONS)
)
export const HCMPlacementRuleList = createResourceReducer(
  resourceReducerFunction,
  predicate.bind(null, RESOURCE_TYPES.HCM_PLACEMENT_RULES)
)
export const userInfoList = createResourceReducer(
  resourceReducerFunction,
  predicate.bind(null, RESOURCE_TYPES.USER_INFO)
)

export { topology } from './topology'
export { resourceFilters } from './filter'
