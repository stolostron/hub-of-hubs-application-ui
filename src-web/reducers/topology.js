/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
// Copyright (c) 2020 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project

import lodash from 'lodash'
import * as Actions from '../actions'
import {
  getFilterState,
  saveFilterState
} from '../../lib/client/filter-helper'
import { RESOURCE_TYPES } from '../../lib/shared/constants'

const initialState = {
  availableFilters: {
    clusters: [],
    labels: [],
    namespaces: [],
    types: []
  },
  activeFilters: {
    namespace: [],
    type: [{ label: 'deployment' }] // Sets the default filters
  },
  diagramFilters: [],
  otherTypeFilters: [],
  links: [],
  nodes: [],
  status: Actions.REQUEST_STATUS.INCEPTION
}

//default k8 type filters
export const defaultTypeFilters = new Set([
  'internet',
  'host',
  'service',
  'deployment',
  'daemonset',
  'statefulset',
  'pod',
  'container'
])

export const topology = (state = initialState, action) => {
  if (
    action.resourceType &&
    action.resourceType.name === RESOURCE_TYPES.HCM_TOPOLOGY.name
  ) {
    switch (action.type) {
    case Actions.RESOURCE_REQUEST: {
      return {
        ...state,
        status: Actions.REQUEST_STATUS.IN_PROGRESS,
        fetchFilters: action.fetchFilters,
        reloading: action.reloading,
        loaded: action.reloading,
        detailsReloading: false
      }
    }
    case Actions.RESOURCE_RECEIVE_SUCCESS: {
      const { links, nodes, willLoadDetails } = action
      return {
        ...state,
        status: Actions.REQUEST_STATUS.DONE,
        nodes,
        links,
        activeFilters: action.fetchFilters,
        loaded: true,
        reloading: false,
        detailsLoaded: !willLoadDetails,
        willLoadDetails
      }
    }
    case Actions.RESOURCE_DETAILS_REQUEST: {
      return {
        ...state,
        detailsStatus: Actions.REQUEST_STATUS.IN_PROGRESS,
        fetchFilters: action.fetchFilters,
        detailsReloading: action.reloading,
        detailsLoaded: action.reloading,
        willLoadDetails: false
      }
    }
    case Actions.RESOURCE_DETAILS_RECEIVE_SUCCESS: {
      const { pods } = action
      return {
        ...state,
        detailsStatus: Actions.REQUEST_STATUS.DONE,
        pods,
        activeFilters: action.fetchFilters,
        detailsLoaded: true,
        detailsReloading: false
      }
    }
    case Actions.RESOURCE_RECEIVE_FAILURE: {
      return {
        ...state,
        status: Actions.REQUEST_STATUS.ERROR,
        fetchError: action.err,
        nodes: action.nodes,
        links: action.links
      }
    }
    case Actions.RESOURCE_RECEIVE_CLUSTER_OFFLINE: {
      return {
        ...state,
        status: Actions.REQUEST_STATUS.CLUSTER_OFFLINE,
        err: action.err,
        detailsLoaded: true,
        loaded: true,
        nodes: [],
        links: []
      }
    }
    }
  }

  switch (action.type) {
  case '@@INIT': {
    return initialState
  }
  case Actions.TOPOLOGY_FILTERS_REQUEST: {
    return {
      ...state,
      filtersStatus: Actions.REQUEST_STATUS.IN_PROGRESS
    }
  }
  case Actions.TOPOLOGY_FILTERS_RECEIVE_ERROR: {
    return {
      ...state,
      filtersStatus: Actions.REQUEST_STATUS.ERROR,
      err: action.err
    }
  }
  case Actions.TOPOLOGY_RESTORE_SAVED_FILTERS: {
    const { filters: activeFilters, otherTypeFilters } = getFilterState(
      initialState.activeFilters
    )
    return { ...state, activeFilters, otherTypeFilters, savingFilters: true }
  }
  case Actions.TOPOLOGY_FILTERS_UPDATE: {
    const activeFilters = { ...state.activeFilters } || {}
    activeFilters[action.filterType] = action.filters
    if (state.savingFilters) {
      const { namespace, name } = action
      saveFilterState(namespace, name, {
        filters: activeFilters,
        otherTypeFilters: state.otherTypeFilters
      })
    }
    return { ...state, activeFilters }
  }
  case Actions.TOPOLOGY_SET_ACTIVE_FILTERS: {
    const { activeFilters, reloading = false } = action
    return {
      ...state,
      status: Actions.REQUEST_STATUS.INCEPTION,
      activeFilters,
      reloading
    }
  }
  case Actions.TOPOLOGY_NAME_SEARCH: {
    const { searchName } = action
    return { ...state, searchName }
  }
  case Actions.TOPOLOGY_FILTERS_RECEIVE_SUCCESS: {
    // The 'clusters' filter is different from other filters.
    // Here we are building the filter options using the cluster labels. When a filter is
    // is selected, we have to use the clusters associated with the label (filterValues).
    const clusterFilters = []
    action.clusters.forEach(({ metadata: c }) => {
      clusterFilters.push({
        label: `name: ${c.name}`,
        filterValues: [c.name]
      })
      Object.keys(c.labels).forEach(labelKey => {
        const existingLabel = clusterFilters.find(
          l => l.label === `${labelKey}: ${c.labels[labelKey]}`
        )
        if (existingLabel) {
          existingLabel.filterValues.push(c.name)
        } else {
          clusterFilters.push({
            label: `${labelKey}: ${c.labels[labelKey]}`,
            filterValues: [c.name]
          })
        }
      })
    })

    // consolidate misc types into "other" type
    const otherTypeFilters = []
    let types = action.types.map(i => ({ label: i }))
    if (types.length > 0) {
      types = types.filter(({ label }) => {
        if (defaultTypeFilters.has(label)) {
          return true
        } else {
          otherTypeFilters.push(label)
          return false
        }
      })
      types.push({ label: 'other' })
    }

    const filters = {
      clusters: clusterFilters,
      labels: action.labels.map(l => ({
        label: `${l.name}: ${l.value}`,
        name: l.name,
        value: l.value
      })),
      namespaces: lodash
        .uniqBy(action.namespaces, 'metadata.name')
        .map(n => ({ label: n.metadata.name })),
      types
    }
    return {
      ...state,
      availableFilters: filters,
      otherTypeFilters,
      filtersStatus: Actions.REQUEST_STATUS.DONE
    }
  }
  case Actions.DIAGRAM_RESTORE_FILTERS: {
    const { namespace, name, initialDiagramFilters } = action
    const { filters: diagramFilters } = getFilterState(
      initialDiagramFilters,
      namespace,
      name
    )
    state.diagramFilters = diagramFilters
    return { ...state }
  }
  case Actions.DIAGRAM_SAVE_FILTERS: {
    const { namespace, name, diagramFilters } = action
    saveFilterState(namespace, name, { filters: diagramFilters })
    state.diagramFilters = diagramFilters
    return { ...state }
  }
  default:
    return { ...state }
  }
}
