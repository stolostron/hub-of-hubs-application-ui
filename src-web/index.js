/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2017, 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
// Copyright (c) 2020 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
import React from 'react'
import { render } from 'react-dom'
import { Provider } from 'react-redux'
import thunkMiddleware from 'redux-thunk'
import { createLogger } from 'redux-logger'
import { createStore, combineReducers, applyMiddleware, compose } from 'redux'
import { ApolloProvider } from 'react-apollo'
import App from './containers/App'
import ScrollToTop from './components/common/ScrollToTop'
import * as reducers from './reducers'
import config from '../lib/shared/config'
import apolloClient from '../lib/client/apollo-client'

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose

const loggerMiddleware = createLogger()

const middleware = [thunkMiddleware] // lets us dispatch() functions
if (
  !window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ &&
  config['featureFlags:reduxLogger']
) {
  middleware.push(loggerMiddleware) // middleware that logs actions
}

// Create Redux store with initial state
const store = createStore(
  combineReducers(reducers),
  composeEnhancers(applyMiddleware(...middleware))
)
import { BrowserRouter } from 'react-router-dom'

render(
  <ApolloProvider client={apolloClient.getClient()}>
    <Provider store={store}>
      <BrowserRouter>
        <ScrollToTop>
          <App url={window.location.pathname} />
        </ScrollToTop>
      </BrowserRouter>
    </Provider>
  </ApolloProvider>,
  document.getElementById('page')
)
