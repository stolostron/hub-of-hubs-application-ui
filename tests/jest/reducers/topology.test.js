/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
// Copyright (c) 2020 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
"use strict";

import { topology } from "../../../src-web/reducers/topology";
import { RESOURCE_TYPES } from "../../../lib/shared/constants";
import * as Actions from "../../../src-web/actions";

describe("topology reducer with topology name", () => {
  it("should return a state with IN_PROGRESS status", () => {
    const state = {
      test: "test"
    };
    const action = {
      resourceType: {
        name: RESOURCE_TYPES.HCM_TOPOLOGY.name
      },
      type: Actions.RESOURCE_REQUEST
    };
    expect(topology(state, action)).toMatchSnapshot();
  });

  it("should return a state with DONE status", () => {
    const state = {
      test: "test",
      activeFilters: "noApplication"
    };
    const action = {
      resourceType: {
        name: RESOURCE_TYPES.HCM_TOPOLOGY.name
      },
      type: Actions.RESOURCE_RECEIVE_SUCCESS,
      fetchFilters: "receivedApplication"
    };
    expect(topology(state, action)).toMatchSnapshot();
  });

  it("should return a state with ERROR status", () => {
    const state = {
      test: "test"
    };
    const action = {
      resourceType: {
        name: RESOURCE_TYPES.HCM_TOPOLOGY.name
      },
      type: Actions.RESOURCE_RECEIVE_FAILURE
    };
    expect(topology(state, action)).toMatchSnapshot();
  });
});

describe("topology reducer", () => {
  it("should return a state with IN_PROGRESS status", () => {
    const state = {
      test: "test"
    };
    const action = {
      resourceType: {
        name: RESOURCE_TYPES.HCM_TOPOLOGY.name
      },
      type: Actions.TOPOLOGY_FILTERS_REQUEST
    };
    expect(topology(state, action)).toMatchSnapshot();
  });

  it("should return a state without status", () => {
    const state = {
      test: "test"
    };
    const action = {
      resourceType: {
        name: RESOURCE_TYPES.HCM_TOPOLOGY.name
      },
      type: Actions.REQUEST_STATUS.ERROR
    };
    expect(topology(state, action)).toMatchSnapshot();
  });

  it("should return a state with ERROR status", () => {
    const state = {
      test: "test"
    };
    const action = {
      resourceType: {
        name: RESOURCE_TYPES.HCM_TOPOLOGY.name
      },
      type: Actions.TOPOLOGY_FILTERS_RECEIVE_SUCCESS,
      clusters: [
        {
          metadata: {
            name: "myminikube",
            labels: {
              clusterip: "9.42.23.217",
              location: "toronto",
              provider: "IBM",
              purpose: "test",
              runtime: "kubernetes"
            }
          },
          __typename: "Cluster"
        }
      ],
      namespaces: [
        {
          metadata: {
            name: "default"
          },
          __typename: "Namespace"
        }
      ],
      labels: [
        {
          name: "app",
          value: "loyalty-level",
          __typename: "Label"
        },
        {
          name: "solution",
          value: "stock-trader",
          __typename: "Label"
        }
      ],
      types: [
        "deployment",
        "host",
        "service",
        "pod",
        "container",
        "daemonset",
        "statefulset"
      ]
    };
    expect(topology(state, action)).toMatchSnapshot();
  });

  it("should return a state without status", () => {
    const state = {
      test: "test"
    };
    const action = {
      resourceType: {
        name: RESOURCE_TYPES.HCM_TOPOLOGY.name
      },
      type: Actions.TOPOLOGY_FILTERS_UPDATE
    };
    expect(topology(state, action)).toMatchSnapshot();
  });
});
