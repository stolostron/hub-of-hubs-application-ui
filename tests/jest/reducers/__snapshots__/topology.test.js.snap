// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`topology reducer should return a state with ERROR status 1`] = `
Object {
  "availableFilters": Object {
    "clusters": Array [
      Object {
        "filterValues": Array [
          "myminikube",
        ],
        "label": "name: myminikube",
      },
      Object {
        "filterValues": Array [
          "myminikube",
        ],
        "label": "clusterip: 9.42.23.217",
      },
      Object {
        "filterValues": Array [
          "myminikube",
        ],
        "label": "location: toronto",
      },
      Object {
        "filterValues": Array [
          "myminikube",
        ],
        "label": "provider: IBM",
      },
      Object {
        "filterValues": Array [
          "myminikube",
        ],
        "label": "purpose: test",
      },
      Object {
        "filterValues": Array [
          "myminikube",
        ],
        "label": "runtime: kubernetes",
      },
    ],
    "labels": Array [
      Object {
        "label": "app: loyalty-level",
        "name": "app",
        "value": "loyalty-level",
      },
      Object {
        "label": "solution: stock-trader",
        "name": "solution",
        "value": "stock-trader",
      },
    ],
    "namespaces": Array [
      Object {
        "label": "default",
      },
    ],
    "types": Array [
      Object {
        "label": "deployment",
      },
      Object {
        "label": "host",
      },
      Object {
        "label": "service",
      },
      Object {
        "label": "pod",
      },
      Object {
        "label": "container",
      },
      Object {
        "label": "daemonset",
      },
      Object {
        "label": "statefulset",
      },
      Object {
        "label": "other",
      },
    ],
  },
  "filtersStatus": "DONE",
  "otherTypeFilters": Array [],
  "test": "test",
}
`;

exports[`topology reducer should return a state with IN_PROGRESS status 1`] = `
Object {
  "filtersStatus": "IN_PROGRESS",
  "test": "test",
}
`;

exports[`topology reducer should return a state without status 1`] = `
Object {
  "test": "test",
}
`;

exports[`topology reducer should return a state without status 2`] = `
Object {
  "activeFilters": Object {
    "undefined": undefined,
  },
  "test": "test",
}
`;

exports[`topology reducer with topology name should return a state with DONE status 1`] = `
Object {
  "activeFilters": "receivedApplication",
  "detailsLoaded": true,
  "links": undefined,
  "loaded": true,
  "nodes": undefined,
  "reloading": false,
  "status": "DONE",
  "test": "test",
  "willLoadDetails": undefined,
}
`;

exports[`topology reducer with topology name should return a state with ERROR status 1`] = `
Object {
  "fetchError": undefined,
  "links": undefined,
  "nodes": undefined,
  "status": "ERROR",
  "test": "test",
}
`;

exports[`topology reducer with topology name should return a state with IN_PROGRESS status 1`] = `
Object {
  "detailsReloading": false,
  "fetchFilters": undefined,
  "loaded": undefined,
  "reloading": undefined,
  "status": "IN_PROGRESS",
  "test": "test",
}
`;
